pkg/catalog/endpoint.go

```diff
@@ -18,3 +20,37 @@ func (mc *MeshCatalog) ListEndpointsForService(svc service.MeshService) ([]endpo
 	}
 	return endpoints, nil
 }
+
+// GetServiceEndpoints returns the highest abstract set of endpoint destinations where the service is made available at.
+// If no LB/virtual IPs are assigned to the service, GetServiceEndpoints will return ListEndpointsForService
+func (mc *MeshCatalog) GetServiceEndpoints(svc service.MeshService) ([]endpoint.Endpoint, error) {
```

Comment:
Is this something that should work the same for any `MeshCataloger` implementation? If so, does it make sense to make this a function that takes a `MeshCataloger` as a parameter vs. implementing this as a method on the struct since it only calls methods that are already part of the interface? 

i.e.:
```go
func GetServiceEndpoints(mc MeshCataloger, svc service.MeshService) ([]endpoint.Endpoint, error) {
```

That way we don't have to add another method to the interface and other implementations could take advantage of this function.

Reply:
tl'dr: providers don't have easy access to service's data, this is an easy-to-fix implementation that can go away as soon as providers can get services, which we'd propose as a refactor that can happen parallely while this commit unblocks a lot of other work that depends on it.

---

what you said is 100% correct. Your suggestions are our original plan, we did reason all that, we however saw some problems with the implementation that are trying to circumvent. 

Originally the implementation of this api should be offloaded to endpoint provider, as having cluster/virtual/proxy/lb/etc.... IP depends on the provider of the service, and the current implementation is leaking k8s stuff into the catalog.

However, in the case of K8s provider, the provider would either (1) require access to the services cache (one under SMIspec right now)  or (2) have it's own service cache or (3) issue API call to get service details (to get the Cluster IP).

More or less, we deemeed that since:
- Endpoint providers do not have easy access to the available service cache under SMIspec, unless we refactor and introduce new shared constructs not currently present (that's some major refactor probs)
- Creating a new service cache for endpoint providers is far from optimal
- issues a new API call for each service seems a no-go (even though to some degree, this might already happen)
and
- the only actual provider currently working and being offered is K8s

we propose this implementations which leaks some k8s in the catalog, but is deffinitely easy to clean up and pushed down to providers for them to implement they can query a services cache.